### DISCUSSION TITLE
beforeEach possible bug

### DISCUSSION THREAD

**Discussion by ryanabaker** [cordejs/corde#1390] [Q&A] [answered]
First I would like to say that I think this library is a fantastic idea!! I am just having trouble with one thing.... I have the following code: 

```
beforeEach(async () => {
  await sleep(3000);
});

test("Invalid command retuns correct message", () => {
  expect("blah").toReturn("You need to enter a valid command!");
});

test("Can NOT request to play a song when NOT inside a voice channel", async () => {
  expect("play dua lipa").toReturn("You have to be in a voice channel to queue a song!");

});

test("Can request a song when inside a voice channel and no music currently going", async () => {
  await bot.joinVoiceChannel("890343128794533922")
  expect("play dua lipa").toReturn(":musical_note: **Searching** :mag_right: `dua lipa`");
});
```

I am attempting to make each test wait a little while before execution, however I receive the following output which suggests that the tests do not follow the `beforeEach`...

```
Tests:      2 passed and 1 failed. 3 total
Time:       4.112s
```

Given that there are 3 tests, and each one should be waiting 3s before execution, this doesn't add up.

Am I doing something wrong, or is this broken?

I have tried grouping them, un-grouping them, adding the `beforeEach` in the group etc...

 I noticed [this link](https://cordejs.org/docs/hooks/) to documentation which seems to suggest that this might not be fully implemented?

Any help is much appreciated!

**Comment by lucasgmagalhaes** [comment, marked as answer]
Hi @ryanabaker, Thank you so much that you enjoyed the lib :). And about your problem, it seems that there is a bug indeed. 
You are not doing anything wrong. I will take a look later.
ooh, one thing, try to put all your tests and the hook inside a `group` clause to see if gonna work.
Nevertheless, I will look after this because it should have work anyway

**Reply by ryanabaker** [reply]
I attempted to put all the tests inside a `group` clause and it does not work. 

When grouping them, I tried putting the `beforeEach` inside the `group` and outside the `group`, both which didn't achieve what I was looking for.

Is the goal that any `beforeEach` at the same scope/level as the `group/test` would execute? 

For example the following code:
```
beforeEach(async () => {
  console.log("THIS IS FROM THE OUTSIDE");
});

group("random group...", () => {
  beforeEach(async () => {
    console.log("THIS IS FROM THE INSIDE");
  });

  test("random test 1", () => {
    console.log("Hello from random test 1")
  });

  test("random test 2", () => {
    console.log("Hello from random test 2")
  });
});

test("random test 3", () => {
  console.log("Hello from random test 3")
});

test("random test 4", async () => {
    console.log("Hello from random test 4")
});
```
Would produce the following messages:
```
THIS IS FROM THE OUTSIDE
THIS IS FROM THE INSIDE
Hello from random test 1
THIS IS FROM THE INSIDE
Hello from random test 2
THIS IS FROM THE OUTSIDE
Hello from random test 3
THIS IS FROM THE OUTSIDE
Hello from random test 3
```

I am happy to run code snippets later if it will help you debug.

Side note, what would be the easiest way to start contributing to this project/are you even looking for help?

**Reply by lucasgmagalhaes** [reply]
Well, fortunately, this bug was fixed in the current [branch](https://github.com/cordejs/corde/tree/expect) that I'm working on. Unfortunately, it has not been released yet. It's missing finish e2e tests and add documentation, so there's some effort to be done kkk

And about your question about contributing. First of all, you have no idea how happy I'm to know that you are interested in contributing to the project <3.

And answering your question. Uhmm, I'm working on version v5 of corde. which you can see some of the changes [here](https://github.com/cordejs/corde/pull/1024). You can fork and get the branch that I'm and may do:

* Increase code coverage
* Refactor functions
* Fix typos
* Improve documentation of functions and/or corde's site
* Add new test functions for corde integration testing library or unity testing library

If you have any suggestions or questions about how to configure corde or any other thing, please let me know.

Cheers!
